source/extensions/access_loggers/grpc/grpc_access_log_impl.h

```diff
@@ -17,24 +19,191 @@ namespace Extensions {
+    void onCreateInitialMetadata(Http::RequestHeaderMap&) override {}
+    void onReceiveInitialMetadata(Http::ResponseHeaderMapPtr&&) override {}
+    void onReceiveMessage(std::unique_ptr<ResponseType>&& message) override {
+      const auto& id = message->id();
+
+      if (parent_.buffered_messages_.find(id) != parent_.buffered_messages_.end()) {
+        // After response wait time exceeded, the state should be Initial.
+        if (parent_.buffered_messages_.at(id).state_ != BufferedMessage::State::Pending) {
+          return;
+        }
+
+        switch (message->status()) {
+        case envoy::service::accesslog::v3::BufferedCriticalAccessLogsResponse::ACK:
+          parent_.stats_.critical_logs_succeeded_.inc();
+          parent_.stats_.pending_critical_logs_.dec();
+          parent_.current_critical_buffer_size_bytes_ -=
+              parent_.buffered_messages_.at(id).message_.ByteSizeLong();
+          parent_.buffered_messages_.erase(id);
+          break;
+        case envoy::service::accesslog::v3::BufferedCriticalAccessLogsResponse::NACK:
+          parent_.buffered_messages_.at(id).state_ = BufferedMessage::State::Initial;
+          break;
+        default:
+          return;
+        }
+
+        return;
+      }
+    }
+    void onReceiveTrailingMetadata(Http::ResponseTrailerMapPtr&&) override {}
+    void onRemoteClose(Grpc::Status::GrpcStatus, const std::string&) override {
+      ASSERT(parent_.active_stream_ != nullptr);
+      if (parent_.active_stream_->stream_ != nullptr) {
+        parent_.active_stream_.reset();
+      }
+    }
+
+    CriticalAccessLoggerGrpcClientImpl<RequestType>& parent_;
+    Grpc::AsyncStream<RequestType> stream_;
+  };
+
+  // Copy messages in the buffer. Take care about memory pressure.
+  void flush(RequestType message) override {
+    if (!active_stream_) {
+      active_stream_ = std::make_unique<ActiveStream>(*this);
+    }
+
+    if (active_stream_->stream_ == nullptr) {
+      active_stream_->stream_ =
+          client_.start(service_method_, *active_stream_, Http::AsyncClient::StreamOptions());
+    }
+
+    if (active_stream_->stream_ == nullptr ||
+        active_stream_->stream_.isAboveWriteBufferHighWatermark()) {
+      active_stream_.reset();
+      return;
+    }
+
+    uint32_t id = MessageUtil::hash(message);
+    const auto message_byte_size = message.ByteSizeLong();
+
+    if (current_critical_buffer_size_bytes_ + message_byte_size > max_critical_buffer_size_bytes_) {
+      stats_.ciritcal_logs_disposed_.inc();
+      return;
+    }
+
+    buffered_messages_[id] = BufferedMessage{nullptr, BufferedMessage::State::Initial, message};
+    current_critical_buffer_size_bytes_ += message_byte_size;
+    stats_.pending_critical_logs_.inc();
+
+    for (auto&& buffered_message : buffered_messages_) {
+      uint32_t id = buffered_message.first;
+      buffered_message.second.message_.set_id(id);
+      buffered_message.second.state_ = BufferedMessage::State::Pending;
```

Comment:
In this case, we should distinguish between `Initial` which means that "just buffered", and pending which means the message is inflight. This is because we switch state here.